### ISSUE TITLE
How to ensure charm apt dependencies

### ISSUE THREAD

**Issue by stub42** [canonical/operator#156]
Before a charm does 'import yaml' or 'import psycopg2', we need to ensure that these non-core dependencies are installed on the host system.

Currently, it looks like we need a stanza like this at the start of src/charm.py:

```python
import subprocess
try:
    import yaml
    import psycopg2
except ImportError:
    subprocess.check_call([
        'apt-get', '--assume-yes', '--option=Dpkg::Options::=--force-confold', 'install',
        'python3-yaml', 'python3-psycopg2',
    ])
    import yaml
    import psycopg2
```

I don't think we can hide this uglyness in hooks/install, as it is expected to be a symlink.

It would be best if this could somehow be bootstrapped using code in the operator framework, so that the 'apt-get installs' get run in a consistent manner with any OS release specific settings.

Perhaps the operator framework could detect at import time that it is a) being imported from a script in hooks/* b) in a hook environment, and if so run platform specific bootstrapping to install apt packages listed in a .txt or .yaml file. Extra points if we can support PPAs, but they are much more likely to contain application dependencies that the charm should be installing, rather than charm dependencies needed for the charm code to run.

**Comment by chipaca** [comment]
We're yet to discuss this (it's on the agenda). If you see us talking about the "runtime", it's about this family of problems.

**Comment by niemeyer** [comment]
As mentioned inline above to make sure people would not just copy & paste, please don't do that as it's a very ugly hack to have such an operation at import time in a Python module. We're working on much better ways to handle this, but meanwhile it would be good enough to do that operation on the on.install event.

**Comment by stub42** [comment]
Agreed, package installation is better done in the install hook handler. This approach should only be used when the charm code itself needs access to the library. Whatever approach is used to bootstrap apt packages for the "runtime" needs to ensure packages are installed in the install and upgrade-charm hooks before the charm code is imported.

**Comment by stub42** [comment]
I've also been told that if the operator pod is destroyed and respun, the apt package dependencies need to be reinstalled.

**Comment by jameinel** [comment]
Actually this is something we need to be careful with, because we probably
also need to recreate the symlinks if we don't have any stable storage.
Though switching to Dispatch is a clearer path forward there.
But it does make sense that a 'stateless' pod that gets restarted is going
to need to redo its initialization.

It feels a lot like we want an explicit event in the operator framework for
'initialized', which gets signaled again if you end up on a different pod.

**Comment by glower** [comment]
Hi,
I am currently trying to migrate an existing charm to the operator framework. We have some additional packages there currently installed with `charmhelpers.fetch.apt_install` and `install_remote`. Do we have a preferred way to do this with operator framework?

**Comment by jameinel** [comment]
We're currently discussing nice ways to bring in external code into your
charm and ensure that it is properly set up for runtime.
Right now we're looking at some sort of 'charm build' step that would
package up your dependencies (like charmhelpers.fetch.apt) and make sure
that you have a proper venv, etc.
We just landed support for Juju 2.8's 'dispatch' hook, which is where we
would likely implement the code for installing things that your *charm*
needs to run.

We don't currently have a wrapper around Apt like charmhelpers, and that is
certainly the sort of thing that we would like to have as a component that
you can bring in for your charm. (for things that you need to install for
the application to run)a

**Comment by mthaddon** [comment]
I think this is addressed by charmcraft's use of requirements.txt isn't it?

**Comment by chipaca** [comment]
not really, because you could have some apt dependencies that are not satisfiable via requirements.txt (e.g. nvidia drivers if you're using cuda)

**Comment by mthaddon** [comment]
Ah indeed, I should learn to read the title of issues before replying to them (**apt** dependencies).

**Comment by petevg** [comment]
This issue is being actively worked on this sprint, via a library.
